File: a-java/ac/T785.java
package ac;

import java.util.Scanner;

/**
 * 快速排序 https://www.acwing.com/problem/content/787/
 * 1. (l + r) >> 1，中间值取左边，最后一步取到的是j，所以分区是(l, j)和(j + 1, r)
 * 2. (l + r + 1) >> 1，中间值取右边，最后一步取到的是i，所以分区是(l, i - 1)和(i, r)
 */
class T785 {

    static void swap(int[] q, int i, int j){
        int t = q[i];
        q[i] = q[j];
        q[j] = t;
    }
    
    static void quick_sort(int[] q, int l, int r) {
        if (l >= r) return;
        
        int i = l - 1;
        int j = r + 1;
        int x = q[(l + r) >> 1];
        
        while (i < j) {
            while (q[++i] < x);
            while (q[--j] > x);
            if (i < j) swap(q, i, j);
        }
        
        quick_sort(q, l, j);
        quick_sort(q, j + 1, r);
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] q = new int[n];
        for (int i = 0; i < n; i++) {
            q[i] = sc.nextInt();
        }
        quick_sort(q, 0, n - 1);
        for (int i = 0; i < n; i++) {
            System.out.print(q[i] + " ");
        }
        sc.close();
    }
}

File: b-go/ac/T785.go
package ac

import "fmt"

// 快速排序 https://www.acwing.com/problem/content/787/
// 1. (l + r) >> 1，中间值取左边，最后一步取到的是j，所以分区是(l, j)和(j + 1, r)
// 2. (l + r + 1) >> 1，中间值取右边，最后一步取到的是i，所以分区是(l, i - 1)和(i, r)
func quickSort(q []int, l, r int) {
    if l >= r {
        return
    }
    
    i, j, x := l - 1, r + 1, q[(l + r) >> 1]
    for i < j {
        for {
            i++
            if q[i] >= x {
                break
            }
        }
        for {
            j--
            if q[j] <= x {
                break
            }
        }
        if i < j{
            q[i], q[j] = q[j], q[i]
        }
    }
    
    quickSort(q, l, j)
    quickSort(q, j + 1, r)
}

func T785() {
	var n int
	fmt.Scanf("%d", &n)
	q := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%d", &q[i])
	}
	quickSort(q, 0, n - 1)
	for i := 0; i < n; i++ {
		fmt.Printf("%d ", q[i])
	}
}
